README.md
![Grunt JSON Schema Faker logo](logo/grunt-jsf-logo.png)

[![grunt-jsonschema-faker npm module](https://nodei.co/npm/grunt-jsonschema-faker.png?downloads=true&stars=true "grunt-jsonschema-faker npm module")](https://www.npmjs.com/package/grunt-jsonschema-faker)

[![NPM version](https://badge.fury.io/js/grunt-jsonschema-faker.png)](http://badge.fury.io/js/grunt-jsonschema-faker)
[![Dependency Status](https://david-dm.org/json-schema-faker/grunt-jsonschema-faker/status.svg)](https://david-dm.org/json-schema-faker/grunt-jsonschema-faker)
[![devDependency Status](https://david-dm.org/json-schema-faker/grunt-jsonschema-faker/dev-status.svg)](https://david-dm.org/json-schema-faker/grunt-jsonschema-faker#info=devDependencies)

# grunt-jsonschema-faker v0.2.7 [![Build Status: Linux](https://travis-ci.org/json-schema-faker/grunt-jsonschema-faker.svg?branch=master)](https://travis-ci.org/json-schema-faker/grunt-jsonschema-faker)

> Grunt task generating fake data according to JSON schema



## Getting Started

If you haven't used [Grunt](http://gruntjs.com/) before, be sure to check out the [Getting Started](http://gruntjs.com/getting-started) guide, as it explains how to create a [Gruntfile](http://gruntjs.com/sample-gruntfile) as well as install and use Grunt plugins. Once you're familiar with that process, you may install this plugin with this command:

```shell
npm install grunt-jsonschema-faker --save-dev
```

Once the plugin has been installed, it may be enabled inside your Gruntfile with this line of JavaScript:

```js
grunt.loadNpmTasks('grunt-jsonschema-faker');
```

*This plugin was designed to work with Grunt 0.4.x. If you're still using grunt v0.3.x it's strongly recommended that [you upgrade](http://gruntjs.com/upgrading-from-0.3-to-0.4).*


## Jsonschema-faker task
_Run this task with the `grunt jsonschema-faker` command._

Task targets, files and options may be specified according to the grunt [Configuring tasks](http://gruntjs.com/configuring-tasks) guide.

Executes [`json-schema-faker`](https://github.com/json-schema-faker/json-schema-faker)
over a [JSON Schema](http://json-schema.org/) file to produce randomized data
that conforms to the schema. In other words, the schema specifies what might be
the content of a JSON (it specifies content, structure, nesting, types, etc.)
and `json-schema-faker` generates such content, using 3rd party libs beneath
([faker.js](https://github.com/Marak/faker.js),
[chance.js](https://github.com/victorquinn/chancejs) and
[randexp.js](http://fent.github.io/randexp.js/)).

### Options

#### indent
Type: `integer`

If defined, generated JSON data is stringified using given indentation. If
undefined, the output becomes a big one-liner.

#### size
Type: `integer`

Specify how many elements conforming to a given schema shall be created.

#### extend
Type: `function`

If defined, the function may extend `json-schema-faker` dependencies. The
function accepts one argument, being `jsf` object itself on which `.extend`
may be called to alter faker.js and chance.js. The grunt-level (outer extend
function) doesn't have to return anything.

#### references
Type: `files`

If defined, jsf will load defined files and allow to use them as referenced
subschemas. [Grunt's `file.expand`](http://gruntjs.com/api/grunt.file#grunt.file.expand)
used to resolve file paths.

#### external
Type: `array` of `object`s

If defined, grunt will load external sources and let them re-use them in your
generated fake data (local grunt target). This is useful when you want to
generate several fake files in which data is **related**.

###### external source parameters

The `external` option needs to be an array of objects, each of them needs to
have following 4 fields defined:

 * *`name`*: `string`. This is just a name of the external source. It will be used in the JSON schema (as a `json-schema-faker` *custom format*)
 * *`generator`*: `string` [`random`|`cycle`]. Currently, two generators are provided, but you may write your own
 * *`src`*: `string`. Path to the external data source file.
 * *`map`*: `function`. This function will be used by [`_.map`](https://lodash.com/docs#map) to modify each element from the source file. You may return only the *id* of each source element, but you may as well need to return a complex structure to re-use.

Read the following section if you're not familiar with *external sources*.

###### example

Let's analyse a typical use case.

For example, groups of users are generated in `group.json` file and the users
are generated in `user.json`. Each user object should have `groupId` attribute
that points to a group object (which has an `id` attribute) in `group.json`
file. But, since both files are generated using a separate grunt task/target,
how can we generate exactly the same values in both files, `group.json` and
`user.json`, if using a random hash function?

This is where the `external` option comes in. You can mock entire database
keeping consistent relations between files.

###### explanation

 * first, you need to specify the order in which the files are generated. In our example `group.json` comes first, then it's `user.json`, because it needs groups to generate itself. `group.json` file might be whatever you want - it might be generated dynamically as well. The `user.json` defines the `external` option.
 * when `user.json` is about to be generated, the grunt task iterates over `external` definitions. For each external source:
   * it loads the content from `src` file
   * map all elements using given `map` function
   * creates a *closure* that will have access to the mapped collection.  The closure returns single values (which are products of `map`) and the strategy depends on `generator` used.
   * the closure is registered as an ordinary [`json-schema-faker` custom format](https://github.com/json-schema-faker/json-schema-faker#custom-formats) using the `name` 
   * the JSON schema will handle custom `name` format.

### Usage Examples

#### Basic

```javascript
grunt.initConfig({
    jsonschema_faker: {
        options: {
            indent: 2
        },
        target: {
            src: ['path/to/schema/file/**/*.json'],
            dest: 'path/to/dest/file.json',
            options: {
                size: 100
            }
        }
    }
});
```

It is also important to define the JSON schema properly. Use [JSON schema
generator](http://jsonschema.net/#/) along with [`json-schema-faker` online
demo](http://json-schema-faker.js.org/) for faster development.

#### Extending dependencies

You may also want to extend faker.js and/or chance. In order to do that, define
a `extend` option function (for each target or globally):

```javascript
grunt.initConfig({
    jsonschema_faker: {
        target: {
            src: ['path/to/schema/file/**/*.json'],
            dest: 'path/to/dest/file.json',
            options: {
                size: 100,
                extend: function(jsf) {
                    jsf.extend('faker', function(faker){
                        faker.locale = "pl"; // or any other language
                        faker.custom = {
                            statement: function(length) {
                                return faker.name.firstName() + " has " + faker.finance.amount() + " on " + faker.finance.account(length) + ".";
                            }
                        };
                        return faker;
                    });
                }
            }
        }
    }
});
```

#### External Sources

Following example illustrates using external sources to generate relative data:

```javascript
grunt.initConfig({
    jsonschema_faker: {
        externalSources: {
            src: 'schema/external-sources.json',
            dest: 'data/external-sources.json',
            options: {
                external: [{
                    name: 'randomUserId',
                    generator: 'random',
                    src: 'data/faker.json',
                    map: function(element) {
                        return element.id;
                    }
                }, {
                    name: 'cycledSsn',
                    generator: 'cycle',
                    src: 'data/chance.json',
                    map: function(element) {
                        return element.ssn;
                    }
                }]
            }
        }
    }
});
```


## Release History

 * 2016-04-07   v0.3.0   dependency update json-schema-faker@0.3.0
 * 2016-04-06   v0.2.7   dependency update / version bump
 * 2016-02-13   v0.2.6   dependency update / version bump
 * 2016-02-05   v0.2.5   dependency update / version bump
 * 2016-01-28   v0.2.4   dependency update / version bump
 * 2015-12-30   v0.2.3   dependency update / version bump
 * 2015-09-03   v0.2.2   lodash as hard dependency (instead of devDependency)
 * 2015-09-02   v0.2.1   repository name changed (github json-schema-faker organisation created) 0.2.0 version related config error fix
 * 2015-08-15   v0.2.0   dependency update / version bump [object Object]
 * 2015-07-13   v0.1.7   references option supported
 * 2015-06-25   v0.1.6   dependency update / version bump
 * 2015-06-24   v0.1.5   dependency update / version bump
 * 2015-06-07   v0.1.4   Moving ownership (user rename) tkoomzaaskz:ducin.
 * 2015-05-15   v0.1.3   Fixed missing file in v0.1.2.
 * 2015-05-15   v0.1.2   Deprecated. External sources support.
 * 2015-05-14   v0.1.1   Fixed test definitions. Updated json-schema-faker to 0.1.9. Allowing extending faker.js and chance.js, new test case.
 * 2015-05-08   v0.1.0   Initial version. Task working correctly. Testing 5 cases - faker, chance, array, integer and boolean. Travis CI integration.

---

Task submitted by [Tomasz Ducin](http://ducin.it)

*This file was generated on Thu Apr 07 2016 19:21:15.*


[![Bitdeli Badge](https://d2weczhvl823v0.cloudfront.net/json-schema-faker/grunt-jsonschema-faker/trend.png)](https://bitdeli.com/free "Bitdeli Badge")



CHANGELOG
v0.3.0:
  date: 2016-04-07
  changes:
    - dependency update json-schema-faker@0.3.0
v0.2.7:
  date: 2016-04-06
  changes:
    - dependency update / version bump
v0.2.6:
  date: 2016-02-13
  changes:
    - dependency update / version bump
v0.2.5:
  date: 2016-02-05
  changes:
    - dependency update / version bump
v0.2.4:
  date: 2016-01-28
  changes:
    - dependency update / version bump
v0.2.3:
  date: 2015-12-30
  changes:
    - dependency update / version bump
v0.2.2:
  date: 2015-09-03
  changes:
    - lodash as hard dependency (instead of devDependency)
v0.2.1:
  date: 2015-09-02
  changes:
    - repository name changed (github json-schema-faker organisation created)
    - 0.2.0 version related config error fix
v0.2.0:
  date: 2015-08-15
  changes:
    - dependency update / version bump
    - jozsef.kerekes contribution: multiple file conversion, sql-like auto-increment
v0.1.7:
  date: 2015-07-13
  changes:
    - references option supported
v0.1.6:
  date: 2015-06-25
  changes:
    - dependency update / version bump
v0.1.5:
  date: 2015-06-24
  changes:
    - dependency update / version bump
v0.1.4:
  date: 2015-06-07
  changes:
    - Moving ownership (user rename) tkoomzaaskz:ducin.
v0.1.3:
  date: 2015-05-15
  changes:
    - Fixed missing file in v0.1.2.
v0.1.2:
  date: 2015-05-15
  changes:
    - Deprecated.
    - External sources support.
v0.1.1:
  date: 2015-05-14
  changes:
    - Fixed test definitions.
    - Updated json-schema-faker to 0.1.9.
    - Allowing extending faker.js and chance.js, new test case.
v0.1.0:
  date: 2015-05-08
  changes:
    - Initial version.
    - Task working correctly.
    - Testing 5 cases - faker, chance, array, integer and boolean.
    - Travis CI integration.


test/jsonschema_faker_test.js
'use strict';

var grunt = require('grunt');
//var _ = require('lodash');

function readFile(file) {
  var contents = grunt.file.read(file);
  if (process.platform === 'win32') {
    contents = contents.replace(/\r\n/g, '\n');
  }
  return contents;
}

function assertJSON(test, path, size, message) {
  var content = JSON.parse(readFile(path));
  if (size > 1) {
    test.equal(content.length, size, message);
  } else if (size === 1) {
    test.equal(content.length, undefined, message);
  }
}

exports.jsonschema_faker = {
  faker: function(test) {
    test.expect(1);

    assertJSON(test,
      'tmp/faker.json',
      20,
      'Should generate 20 faker elements');

    test.done();
  },
  chance: function(test) {
    test.expect(1);

    assertJSON(test,
      'tmp/chance.json',
      3,
      'Should generate 2 chance elements');

    test.done();
  },
  list: function(test) {
    test.expect(3);
    var isIndexOk = true,
      listItems = JSON.parse(readFile('tmp/list.json')),
      index, 
      startIndex = listItems[0].id;

    test.equal(listItems.length, 5, "The list contains 5 items.");
    test.equal(startIndex, 10, "Index starts from 10.");
    for (index=0; index<listItems.length; index++){
      if (index + startIndex !== listItems[index].id){
        isIndexOk = false;
      }
    }
    test.ok(isIndexOk, "The index has an ascending order.");
    test.done();
  },
  array: function(test) {
    test.expect(1);

    assertJSON(test,
      'tmp/array.json',
      200,
      'Should generate 200 array elements');

    test.done();
  },
  integer: function(test) {
    test.expect(1);

    assertJSON(test,
      'tmp/integer.json',
      1,
      'Should generate 1 integer element');

    test.done();
  },
  boolean: function(test) {
    test.expect(1);

    assertJSON(test,
      'tmp/boolean.json',
      1,
      'Should generate 1 boolean element');

    test.done();
  },
  fakerExtended: function(test) {
    test.expect(1);

    assertJSON(test,
      'tmp/faker-extended.json',
      30,
      'Should generate 30 faker-extended elements');

    test.done();
  },
  externalSources: function(test) {
    test.expect(1);

    assertJSON(test,
      'tmp/external-sources.json',
      15,
      'Should generate 15 elements with external sources re-used');

    test.done();
  }
};
  

generators.js
/**
 * @see https://github.com/json-schema-faker/grunt-jsonschema-faker#external
 *
 * This file contains generator functions that are used as intermediate
 * preprocessors for external data collections loaded into JSON Schema.
 * When a collection, having *n* items, is loaded as external data source, there
 * is a question: which item to choose for particular JSON generated object?
 * The functions below give the opportunity to make it one by one (a cycle),
 * just a random value or anything else.
 */

var _ = require('lodash');

module.exports = {
    random: function(collection){
        return function(){
            return _.sample(collection);
        };
    },
    cycle: function(collection){
        var ind = 0;
        return function(){
            var res = collection[ind];
            ind++;
            if (ind === collection.length){
                ind = 0;
            }
            return res;
        };
    }
};


tasks/jsonschema_faker.js
/**
 * grunt-jsonschema-faker is a Grunt task generating fake data according to JSON schema
 *
 * @see https://github.com/json-schema-faker/grunt-jsonschema-faker
 *
 * @author Tomasz Ducin <tomasz@ducin.it> (https://github.com/ducin)
 * @copyright © 2015-2016 Tomasz Ducin
 * @license MIT https://raw.github.com/json-schema-faker/grunt-jsonschema-faker/blob/master/LICENSE
 */

'use strict';


module.exports = function(grunt){
  var path = require('path');
  var _ = require('lodash');
  var jsf = require('json-schema-faker');
  var externalFormats = require('../formats');
  var externalGenerators = require('../generators');

  grunt.registerMultiTask('jsonschema_faker', 'generating fake data', function(){
    var options = this.options();
    if (options.extend){
        options.extend(jsf);
    }

    // if there are any externals defined, then load generators to handle them
    // (https://github.com/json-schema-faker/grunt-jsonschema-faker#external)
    var references = [];
    if (options.references){
      grunt.file.expand(options.references).forEach(function(f){
        references.push(grunt.file.readJSON(f));
      });
    }

    // load additional JSON-Schema-faker formats
    _.each(externalFormats, function(formatFunction, formatName){
        jsf.format(formatName, formatFunction());
    });

    this.files.forEach(function (f){
      var cwd = path.normalize(f.orig.cwd || ''),
          cwdAbs = path.resolve(cwd || '.'),
          expand = !!f.orig.expand;
      f.src.map(function (file){
          file = path.normalize(file);
          return path.resolve(cwdAbs, (expand && cwd.length && (file.indexOf(cwd + path.sep) === 0)) ? file.substr(cwd.length + path.sep.length) : file);
      }).filter(function (file){
          if (!grunt.file.exists(file)){
              grunt.log.warn('Source file "' + file + '" not found.');
              return false;
          } else {
              grunt.log.ok('Reading file ' + file + '.');
              return true;
          }
      }).map(function(file){
        _.each(options.external, function(externalOptions){
            var externalContent = grunt.file.readJSON(externalOptions.src);
            var collection = _.map(externalContent, externalOptions.map);
            var generator = externalGenerators[externalOptions.generator];
            jsf.format(externalOptions.name, generator(collection));
        });
        var schema = grunt.file.readJSON(file);
        var result = !options.size ? jsf(schema, references) :
            _.map(_.range(options.size), function(){
              return jsf(schema, references);
            });
        var destination = path.normalize(f.dest);
        grunt.log.ok('Writing file ' + destination + '.\n');
        grunt.file.write(destination, JSON.stringify(result, null, options.indent));
      });
    });
  });
};


Gruntfile.js
/*
 * grunt-jsonschema-faker
 * https://github.com/json-schema-faker/grunt-jsonschema-faker
 *
 * Copyright (c) 2015-2016 Tomasz Ducin
 * Licensed under the MIT license.
 */

'use strict';

module.exports = function (grunt){

    // Project configuration.
    grunt.initConfig({
        jsonschema_faker: {
            options: {
                indent: 2
            },
            faker: {
                src: 'test/fixtures/faker.json',
                dest: 'tmp/faker.json',
                options: {
                    size: 20
                }
            },
            chance: {
                src: 'test/fixtures/chance.json',
                dest: 'tmp/chance.json',
                options: {
                    size: 3
                }
            },
            list: {
                src: 'test/fixtures/list.json',
                dest: 'tmp/list.json',
                options: {
                    size: 5 
                }
            },
            array: {
                src: 'test/fixtures/array.json',
                dest: 'tmp/array.json',
                options: {
                    size: 200
                }
            },
            integer: {
                src: 'test/fixtures/integer.json',
                dest: 'tmp/integer.json'
            },
            boolean: {
                src: 'test/fixtures/boolean.json',
                dest: 'tmp/boolean.json'
            },
            multi: {
                files: [
                    {
                        expand: true,
                        cwd: "test/fixtures",
                        src: "?????.json", // 5-letter-long files
                        dest: "tmp/multi1",
                        ext: '.json',
                        rename: function (dest, src){
                            return dest + "/renamed-" + src;
                        }
                    },
                    {
                        expand: true,
                        cwd: "test/fixtures",
                        src: "???????.json", // 7-letter-long files
                        dest: "tmp/multi2",
                        ext: '.json',
                        rename: function (dest, src){
                            return dest + "/renamed-" + src;
                        }
                    }
                ]
            },
            fakerExtended: {
                src: 'test/fixtures/faker-extended.json',
                dest: 'tmp/faker-extended.json',
                options: {
                    size: 30,
                    extend: function(jsf){
                        jsf.extend('faker', function(faker){
                            faker.locale = "pl"; // or any other language
                            faker.custom = {
                                statement: function(length){
                                    return faker.name.firstName() + " has " + faker.finance.amount() + " on " + faker.finance.account(length) + ".";
                                }
                            };
                            return faker;
                        });
                    }
                }
            },
            externalSources: {
                src: 'test/fixtures/external-sources.json',
                dest: 'tmp/external-sources.json',
                options: {
                    size: 15,
                    external: [{
                        name: 'randomUserId',
                        generator: 'random',
                        src: 'tmp/faker.json',
                        map: function(element){
                            return element.id;
                        }
                    }, {
                        name: 'cycledSsn',
                        generator: 'cycle',
                        src: 'tmp/chance.json',
                        map: function(element){
                            return element.ssn;
                        }
                    }]
                }
            }
        },
        jshint: {
            all: [
                'Gruntfile.js',
                'tasks/*.js',
                '<%= nodeunit.tests %>'
            ],
            options: {
                jshintrc: '.jshintrc'
            }
        },

        clean: {
            tmp: ['tmp'],
        },

        // Unit tests.
        nodeunit: {
            tests: ['test/*_test.js'],
        },

        // Post contrib-build.
        concat: {
            'README.md': ['docs/header.md', 'README.md']
        }
    });

    // Actually load this plugin's task(s).
    grunt.loadTasks('tasks');

    // These plugins provide necessary tasks.
    grunt.loadNpmTasks('grunt-contrib-jshint');
    grunt.loadNpmTasks('grunt-contrib-clean');
    grunt.loadNpmTasks('grunt-contrib-nodeunit');
    grunt.loadNpmTasks('grunt-contrib-internal');

    // Whenever the 'test' task is run, first run jshint, clean the placeholder,
    // then run this plugin's task(s), then test the result.
    grunt.registerTask('test', ['jshint', 'clean', 'jsonschema_faker', 'nodeunit']);

    grunt.loadNpmTasks('grunt-contrib-concat');
    grunt.registerTask('build', ['build-contrib', 'concat']);

    // By default, lint and run all tests.
    grunt.registerTask('default', ['jshint', 'test', 'build']);
};
